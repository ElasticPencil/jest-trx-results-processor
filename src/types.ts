export type Epoch = number;

export type JestTestResult = {
  /**
   * Array of messages in the 'describe' blocks surrounding the test.
   * 
   * @type {string[]}
   */
  ancestorTitles: string[];
  /**
   * Array of error messages.
   * 
   * @type {string[]}
   */
  failureMessages: string[];
  /**
   * Total number of asserts that are passing.
   * 
   * @type {number}
   */
  numPassingAsserts: number;
  /**
   * Status of the test.
   * 
   * @type {("failed" | "pending" | "passed")}
   */
  status: "failed" | "pending" | "passed";
  /**
   * Message in the 'it' block.
   * 
   * @type {string}
   */
  title: string;
};

export type JestTestSuiteResult = {
  /**
   * Test coverage information.
   * 
   * @type {{}}
   */
  coverage: {};
  /**
   * Number of failing tests in the suite.
   * 
   * @type {number}
   */
  numFailingTests: number;
  /**
   * Number of passing tests in the suite.
   * 
   * @type {number}
   */
  numPassingTests: number;
  /**
   * Number of pending tests in the suite.
   * 
   * @type {number}
   */
  numPendingTests: number;
  /**
   * Performance information.
   * 
   * @type {{
   *     start: Epoch;
   *     end: Epoch;
   *   }}
   */
  perfStats: {
    /**
     * Unix timestamp of the test suite run start.
     * 
     * @type {Epoch}
     */
    start: Epoch;
    /**
     * Unix timestamp of the test suite run end.
     * 
     * @type {Epoch}
     */
    end: Epoch;
  };
  /**
   * Absolute path to the test file.
   * 
   * @type {string}
   */
  testFilePath: string;
  /**
   * Results of the individual tests.
   * 
   * @type {JestTestResult[]}
   */
  testResults: JestTestResult[];
};

export type JestTestRunResult = {
  /**
   * True if all the tests are passing.
   * 
   * @type {boolean}
   */
  success: boolean;
  /**
   * Unix timestamp of the test run start.
   * 
   * @type {Epoch}
   */
  startTime: Epoch;
  /**
   * Total number of test suites.
   * 
   * @type {number}
   */
  numTotalTestSuites: number;
  /**
   * Number of test suites that have all the tests passing.
   * 
   * @type {number}
   */
  numPassedTestSuites: number;
  /**
   * Number of test suites that have at least one failing test.
   * 
   * @type {number}
   */
  numFailedTestSuites: number;
  /**
   * Number of test suites that ended with a runtime error.
   * 
   * @type {number}
   */
  numRuntimeErrorTestSuites: number;
  /**
   * Total number of tests across all the test suites.
   * 
   * @type {number}
   */
  numTotalTests: number;
  /**
   * Number of passing tests across all the test suites.
   * 
   * @type {number}
   */
  numPassedTests: number;
  /**
   * Number of failing tests across all the test suites.
   * 
   * @type {number}
   */
  numFailedTests: number;
  /**
   * Number of pending tests across all the test suites.
   * 
   * @type {number}
   */
  numPendingTests: number;
  /**
   * Results of individual test suites.
   * 
   * @type {JestTestSuiteResult[]}
   */
  testResults: JestTestSuiteResult[];
};
