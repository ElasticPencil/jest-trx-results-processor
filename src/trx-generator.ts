import * as builder from "xmlbuilder";
import * as uuid from "uuid";
import * as os from "os";

import { JestTestResult, JestTestRunResult, JestTestSuiteResult } from "./types";

// Magic GUID's and other constants
const testListNotInListId = "8c84fa94-04c1-424b-9868-57a2d4851a1d";
const testListAllLoadedResultsId = "19431567-8539-422a-85d7-44ee4e166bda";
const testType = "13cdc9d9-ddb5-4fa4-a97d-d965ccfc6d4b";

const testOutcomeTable: { [outcome: string]: string } = {
  "failed": "Failed",
  "pending": "Skipped",
  "passed": "Passed"
};

// Auxilliary functions
const getFullTestName = (testResult: JestTestResult): string =>
  testResult.ancestorTitles && testResult.ancestorTitles.length
    ? `${testResult.ancestorTitles.join(" / ")} / ${testResult.title}`
    : testResult.title;

const getTestClassName = (testResult: JestTestResult): string =>
  testResult.ancestorTitles && testResult.ancestorTitles.length
    ? testResult.ancestorTitles[0]
    : "No suite";

const getSuitePerTestDuration = (testSuiteResult: JestTestSuiteResult): number =>
  // take the total duration of suite and divide it by the number of tests (Jest does not provide per test performance info)
  Math.floor((testSuiteResult.perfStats.end - testSuiteResult.perfStats.start) / (testSuiteResult.numPassingTests + testSuiteResult.numFailingTests + testSuiteResult.numPendingTests));

// Adapted from https://github.com/hatchteam/karma-trx-reporter
const formatDuration = (duration: number): string => {
  let durationInner = duration | 0;
  const ms = durationInner % 1000;
  durationInner -= ms;
  const s = (durationInner / 1000) % 60;
  durationInner -= s * 1000;
  const m = (durationInner / 60000) % 60;
  durationInner -= m * 60000;
  const h = (durationInner / 3600000) % 24;
  durationInner -= h * 3600000;
  const d = durationInner / 86400000;

  return (d > 0 ? d + "." : "") +
    (h < 10 ? "0" + h : h) + ":" +
    (m < 10 ? "0" + m : m) + ":" +
    (s < 10 ? "0" + s : s) + "." +
    (ms < 10 ? "00" + ms : ms < 100 ? "0" + ms : ms);
};

const sanitizationRegex = /[\u0000-\u0008\u000B-\u000C\u000E-\u001F\uD800-\uDFFF\uFFFE-\uFFFF]/g;
const sanitizeString = (str: string): string => str && str.replace(sanitizationRegex, ""); // removes the characters that make xmlbuilder throw

export const generateTrx = (testRunResult: JestTestRunResult): string => {
  const computerName = os.hostname();
  const userName = process.env.SUDO_USER ||
    process.env.LOGNAME ||
    process.env.USER ||
    process.env.LNAME ||
    process.env.USERNAME;

  // TestRun
  const resultBuilder = builder.create("TestRun", { version: "1.0", encoding: "UTF-8" })
    .att("id", uuid.v4())
    .att("name", `${userName}@${computerName} ${new Date(testRunResult.startTime).toISOString()}`)
    .att("runUser", userName)
    .att("xmlns", "http://microsoft.com/schemas/VisualStudio/TeamTest/2010");

  // TestSettings
  resultBuilder.ele("TestSettings")
    .att("name", "Jest test run")
    .att("id", uuid.v4());

  // Times
  const startTime = new Date(testRunResult.startTime).toISOString();
  resultBuilder.ele("Times")
    .att("creation", startTime)
    .att("queuing", startTime)
    .att("start", startTime);

  // ResultSummary
  const summary = resultBuilder.ele("ResultSummary")
    .att("outcome", testRunResult.success ? "Passed" : "Failed");
  summary.ele("Counters")
    .att("total", testRunResult.numTotalTests)
    .att("executed", testRunResult.numTotalTests - testRunResult.numPendingTests)
    .att("passed", testRunResult.numPassedTests)
    .att("failed", testRunResult.numFailedTests)
    .att("error", 0);

  // TestDefinitions
  const testDefinitions = resultBuilder.ele("TestDefinitions");
  const testLists = resultBuilder.ele("TestLists");
  testLists.ele("TestList")
    .att("name", "Results Not in a List")
    .att("id", testListNotInListId);
  testLists.ele("TestList")
    .att("name", "All Loaded Results")
    .att("id", testListAllLoadedResultsId);
  const testEntries = resultBuilder.ele("TestEntries");
  const results = resultBuilder.ele("Results");

  testRunResult.testResults.forEach(testSuiteResult => {
    const perTestDuration = getSuitePerTestDuration(testSuiteResult);
    const perTestDurationFormatted = formatDuration(perTestDuration);

    testSuiteResult.testResults.forEach((testResult, index) => {
      const testId = uuid.v4();
      const executionId = uuid.v4();
      const fullTestName = getFullTestName(testResult);

      // UnitTest
      const unitTest = testDefinitions.ele("UnitTest")
        .att("name", fullTestName)
        .att("id", testId);
      unitTest.ele("Execution")
        .att("id", executionId);
      unitTest.ele("TestMethod")
        .att("codeBase", `Jest_${fullTestName}`)
        .att("name", fullTestName)
        .att("className", getTestClassName(testResult));

      // TestEntry
      testEntries.ele("TestEntry")
        .att("testId", testId)
        .att("executionId", executionId)
        .att("testListId", testListNotInListId);

      // UnitTestResult
      const result = results.ele("UnitTestResult")
        .att("testId", testId)
        .att("executionId", executionId)
        .att("testName", fullTestName)
        .att("computerName", computerName)
        .att("duration", perTestDurationFormatted)
        .att("startTime", new Date(testSuiteResult.perfStats.start + index * perTestDuration).toISOString())
        .att("endTime", new Date(testSuiteResult.perfStats.start + (index + 1) * perTestDuration).toISOString())
        .att("testType", testType)
        .att("outcome", testOutcomeTable[testResult.status])
        .att("testListId", testListNotInListId);

      if (testResult.status === "failed") {
        result.ele("Output").ele("ErrorInfo").ele("Message", sanitizeString(testResult.failureMessages.join("\n")));
      }
    });
  });

  return resultBuilder.end({ pretty: true });
};
