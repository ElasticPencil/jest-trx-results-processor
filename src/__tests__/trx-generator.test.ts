import "jest";
import { generateTrx } from "../trx-generator";
import { JestTestRunResult } from "../types";

describe("trx-generator", (): void => {
  it("processes the results correctly", (): void => {
    const input: JestTestRunResult = {
      success: true,
      startTime: 1478771929,
      numTotalTestSuites: 1,
      numPassedTestSuites: 0,
      numFailedTestSuites: 1,
      numRuntimeErrorTestSuites: 0,
      numTotalTests: 2,
      numPassedTests: 1,
      numFailedTests: 1,
      numPendingTests: 0,
      testResults: [
        {
          coverage: {},
          numFailingTests: 1,
          numPassingTests: 1,
          numPendingTests: 0,
          perfStats: {
            start: 1478771929,
            end: 1478778929,
          },
          testFilePath: "C:\\testPath\\test.js",
          testResults: [
            {
              ancestorTitles: ["foo's", "bar method"];
              failureMessages: [];
              numPassingAsserts: 1;
              status: "passed";
              title: "works well";
            },
            {
              ancestorTitles: ["foo's", "bar method"];
              failureMessages: ["This did not go as planned"];
              numPassingAsserts: 1;
              status: "failed";
              title: "works not so well";
            }
          ]
        }
      ],
    };
    const result: string = generateTrx(input);
    expect(result).toMatchSnapshot();
  });
});