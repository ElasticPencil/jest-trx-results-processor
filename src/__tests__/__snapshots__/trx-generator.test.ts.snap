exports[`trx-generator processes the results correctly 1`] = `
"<?xml version=\"1.0\" encoding=\"UTF-8\"?>
<TestRun id=\"a4495492-7559-4db1-99a7-3baae89b77cf\" name=\"homod99@N099A0248 1970-01-18T02:46:11.929Z\" runUser=\"homod99\" xmlns=\"http://microsoft.com/schemas/VisualStudio/TeamTest/2010\">
  <TestSettings name=\"Jest test run\" id=\"1d058708-9b4e-4e0b-99b4-215d10822b02\"/>
  <Times creation=\"1970-01-18T02:46:11.929Z\" queuing=\"1970-01-18T02:46:11.929Z\" start=\"1970-01-18T02:46:11.929Z\"/>
  <ResultSummary outcome=\"Passed\">
    <Counters total=\"2\" executed=\"2\" passed=\"1\" failed=\"1\" error=\"0\"/>
  </ResultSummary>
  <TestDefinitions>
    <UnitTest name=\"foo\'s / bar method / works well\" id=\"cf3350a3-a0a6-4933-b624-1dec8ef8e6dd\">
      <Execution id=\"53fb66f0-4c05-4543-9081-da44e373cd99\"/>
      <TestMethod codeBase=\"Jest_foo\'s / bar method / works well\" name=\"foo\'s / bar method / works well\" className=\"foo\'s\"/>
    </UnitTest>
    <UnitTest name=\"foo\'s / bar method / works not so well\" id=\"7c2de6d3-2109-4116-94bf-2ea328b2b02a\">
      <Execution id=\"1e5fa663-c66c-470d-982e-69f26ee500be\"/>
      <TestMethod codeBase=\"Jest_foo\'s / bar method / works not so well\" name=\"foo\'s / bar method / works not so well\" className=\"foo\'s\"/>
    </UnitTest>
  </TestDefinitions>
  <TestLists>
    <TestList name=\"Results Not in a List\" id=\"8c84fa94-04c1-424b-9868-57a2d4851a1d\"/>
    <TestList name=\"All Loaded Results\" id=\"19431567-8539-422a-85d7-44ee4e166bda\"/>
  </TestLists>
  <TestEntries>
    <TestEntry testId=\"cf3350a3-a0a6-4933-b624-1dec8ef8e6dd\" executionId=\"53fb66f0-4c05-4543-9081-da44e373cd99\" testListId=\"8c84fa94-04c1-424b-9868-57a2d4851a1d\"/>
    <TestEntry testId=\"7c2de6d3-2109-4116-94bf-2ea328b2b02a\" executionId=\"1e5fa663-c66c-470d-982e-69f26ee500be\" testListId=\"8c84fa94-04c1-424b-9868-57a2d4851a1d\"/>
  </TestEntries>
  <Results>
    <UnitTestResult testId=\"cf3350a3-a0a6-4933-b624-1dec8ef8e6dd\" executionId=\"53fb66f0-4c05-4543-9081-da44e373cd99\" testName=\"foo\'s / bar method / works well\" computerName=\"N099A0248\" duration=\"00:00:03.500\" startTime=\"1970-01-18T02:46:11.929Z\" endTime=\"1970-01-18T02:46:15.429Z\" testType=\"13cdc9d9-ddb5-4fa4-a97d-d965ccfc6d4b\" outcome=\"Passed\" testListId=\"8c84fa94-04c1-424b-9868-57a2d4851a1d\"/>
    <UnitTestResult testId=\"7c2de6d3-2109-4116-94bf-2ea328b2b02a\" executionId=\"1e5fa663-c66c-470d-982e-69f26ee500be\" testName=\"foo\'s / bar method / works not so well\" computerName=\"N099A0248\" duration=\"00:00:03.500\" startTime=\"1970-01-18T02:46:15.429Z\" endTime=\"1970-01-18T02:46:18.929Z\" testType=\"13cdc9d9-ddb5-4fa4-a97d-d965ccfc6d4b\" outcome=\"Failed\" testListId=\"8c84fa94-04c1-424b-9868-57a2d4851a1d\">
      <Output>
        <ErrorInfo>
          <Message>This did not go as planned</Message>
        </ErrorInfo>
      </Output>
    </UnitTestResult>
  </Results>
</TestRun>"
`;
